src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+
+	shardConnections = (ShardConnections *) hash_search(shardConnectionHash,
+														&shardId,
+														HASH_ENTER,
+														shardConnectionsFound);
+	if (!*shardConnectionsFound)
+	{
+		/* intialize COPY transactions on shard placements */
+		shardConnections->shardId = shardId;
+		shardConnections->connectionList = NIL;
+
+		OpenCopyTransactions(copyStatement, shardConnections);
+	}
+
+	return shardConnections;
+}
+
+
+/*
+ * OpenCopyTransactions opens a connection for each placement of a shard and
+ * starts a COPY transaction. If a connection cannot be opened, then the shard
+ * placement is marked as inactive and the COPY continues with the remaining
+ * shard placements.
+ */
+static void
+OpenCopyTransactions(CopyStmt *copyStatement, ShardConnections *shardConnections)
+{
+	List *finalizedPlacementList = NIL;
+	List *failedPlacementList = NIL;
+	ListCell *placementCell = NULL;
+	ListCell *failedPlacementCell = NULL;
+	List *connectionList = NIL;
+
+	finalizedPlacementList = FinalizedShardPlacementList(shardConnections->shardId);
+
+	foreach(placementCell, finalizedPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		char *nodeName = placement->nodeName;
+		int nodePort = placement->nodePort;
+		TransactionConnection *transactionConnection = NULL;
+		StringInfo copyCommand = NULL;
+		PGresult *result = NULL;
+
+		PGconn *connection = ConnectToNode(nodeName, nodePort);
+		if (connection == NULL)
+		{
+			failedPlacementList = lappend(failedPlacementList, placement);
+			continue;
+		}
+
+		result = PQexec(connection, "BEGIN");
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			ReportRemoteError(connection, result);
+			failedPlacementList = lappend(failedPlacementList, placement);
+			continue;
+		}
+
+		copyCommand = ConstructCopyStatement(copyStatement, shardConnections->shardId);
+
+		result = PQexec(connection, copyCommand->data);
+		if (PQresultStatus(result) != PGRES_COPY_IN)
+		{
+			ReportRemoteError(connection, result);
+			failedPlacementList = lappend(failedPlacementList, placement);
+			continue;
+		}
+
+		transactionConnection = palloc0(sizeof(TransactionConnection));
+
+		transactionConnection->connectionId = shardConnections->shardId;
+		transactionConnection->transactionState = TRANSACTION_STATE_COPY_STARTED;
+		transactionConnection->connection = connection;
+
+		connectionList = lappend(connectionList, transactionConnection);
+	}
+
+	/* if all placements failed, error out */
+	if (list_length(failedPlacementList) == list_length(finalizedPlacementList))
+	{
+		ereport(ERROR, (errmsg("could not find any active placements")));
+	}
+
+	/* otherwise, mark failed placements as inactive: they're stale */
+	foreach(failedPlacementCell, failedPlacementList)
+	{
+		ShardPlacement *failedPlacement = (ShardPlacement *) lfirst(failedPlacementCell);
+		uint64 shardLength = 0;
+
+		DeleteShardPlacementRow(failedPlacement->shardId, failedPlacement->nodeName,
```

Comment:
Our locking around this is, uh, a bit weak. But I don't think this is the only place with that problem.